### PR TITLE
Remove ZeebeRaftStateMachine

### LINKED ISSUE #4914: Remove ZeebeRaftStateMachine
**Description**

A state machine in a raft is generally used to build the state by applying committed events. We built the state differently via stream processors. So it is not necessary to have a different state machine. The current state machine reads and applies every entry that is committed. Here applying is only used to notify commit listeners.

**Problems with current approach**

- On restart, it has to apply all events in the log. This leads to
   - Longer restart time (#4158)
   - Out of memory when log is too large (#4871) (We can fix OOM due to this independently of this issue)
- On normal run, it might have an impact on performance as it has to re-read the entries (on followers).

**Proposal**

We can remove the state machine and the need for "apply"ing the entries. The commit listener is only used by the logstream to update the commit position. Commit notification can be done by the `LeaderRole` when an entry is committed. No need to wait for applying. 

**Cons**
On follower it cannot get notified for each entry individually. As a consequence, it is not possible to update the commit position in logstream reliable. We, currently,  do not update and use the commit position on followers. So this is not an issue now. But it can cause problems in future, if we ever want to use this on the followers.

### PR DESCRIPTION
**Pull request by deepthidevaki**
## Description

* Use commit listener in LogStorageAppender to update logstream commitPosition instead of raft commit listener.
* Raft commit listener now only notifies with the raft index instead of the entry at the index. This way it does not have to read the log to notify listeners.
* Do not "apply" entries after the commit. Applying was only used to notify commit listeners. Rename ZeebeRaftStateMachine to LogCompactor
* Some cleanup of unused code and classes

## Related issues

closes #4914 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to backport the fix to the last two minor versions

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [x] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the release announcement

### CODE REVIEW

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 714b54c07cdf71eb5cf80851843667f05c4a76be
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert, ci, build**

Guidelines are available at https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

**Comment by deepthidevaki** [comment]
Results from running a benchmark. 

image

**Review by ChrisKujawa — CHANGES_REQUESTED** [review]
Thanks @deepthidevaki :+1: 

I have some smaller comments

**Review by miguelpires — CHANGES_REQUESTED** [review]
Looks good, just had a few comments and suggestions

File: atomix/cluster/src/main/java/io/atomix/raft/roles/LeaderRole.java [hunk 1]
```diff
@@ -93,27 +92,6 @@ public LeaderRole(final RaftContext context) {
     return super.start().thenRun(this::startTimers).thenApply(v -> this);
   }
 
-  private ZeebeEntry findLastZeebeEntry() {
```
[the remark this answers is not in the document]
Comment (Zelldon) [reply atomix/cluster/src/main/java/io/atomix/raft/roles/LeaderRole.java:96]: Agree with miguel here. I already pointed this out in other PR's I might due that atomix is not completely formatted? Or does it happen also in other modules?

[the remark this answers is not in the document]
Comment (deepthidevaki) [reply atomix/cluster/src/main/java/io/atomix/raft/roles/LeaderRole.java:96]: Our codestyle says "Overridden methods must be grouped together". I believe that is what this auto-formatting did. What do you suggest here? Should I revert these formatting?

File: atomix/cluster/src/main/java/io/atomix/raft/partition/RaftPartitionGroupConfig.java [hunk 2]
```diff
@@ -39,39 +37,10 @@
   private Duration electionTimeout = DEFAULT_ELECTION_TIMEOUT;
   private Duration heartbeatInterval = DEFAULT_HEARTBEAT_INTERVAL;
   private RaftStorageConfig storageConfig = new RaftStorageConfig();
-  private RaftCompactionConfig compactionConfig = new RaftCompactionConfig();
```
[the remark this answers is not in the document]
Comment (deepthidevaki) [reply atomix/cluster/src/main/java/io/atomix/raft/partition/RaftPartitionGroupConfig.java:42]: As far as I know this was used by the PartitionManagementGroupService, which sends the configurations to other nodes. But we removed it. So now I don't find any place where this config is persisted or used for communication with other nodes. Also our upgradability test and rolling update tests passed. Do you have any other suggestions on how to test that this does not break compatability?

File: broker/src/main/java/io/zeebe/broker/system/partitions/ZeebePartition.java [hunk 3]
```diff
@@ -654,21 +646,6 @@ private void stepByStepClosing(
             });
   }
 
-  @Override
-  public <T extends RaftLogEntry> void onCommit(final Indexed<T> indexed) {
```
[the remark this answers is not in the document]
Comment (deepthidevaki) [reply broker/src/main/java/io/zeebe/broker/system/partitions/ZeebePartition.java:658]: This is removed. This was used only to set the logstream commit position. It is now done by the LogStorageAppender. Downside of this approach is that followers cannot update commit position, which is not required now.

**Comment by deepthidevaki** [comment]
@Zelldon @MiguelPires I have updated the PR - removed "lastApplied" and use commitIndex instead because both are same now. Also see my comments regarding backward compatability and reformatting.

File: atomix/cluster/src/main/java/io/atomix/raft/partition/RaftPartitionGroupConfig.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Zelldon) [reply atomix/cluster/src/main/java/io/atomix/raft/partition/RaftPartitionGroupConfig.java:42]: Ok thanks. Was not sure where this is done to be honest.

File: atomix/cluster/src/main/java/io/atomix/raft/impl/RaftContext.java [hunk 4]
```diff
@@ -723,28 +727,6 @@ public void setEntryValidator(final EntryValidator validator) {
...
-    return lastApplied;
-  }
-
-  /**
-   * Sets the last applied index.
-   *
-   * @param lastApplied the last applied index and the last applied term
-   */
-  private void setLastApplied(final long lastApplied) {
-    this.lastApplied = Math.max(this.lastApplied, lastApplied);
```
Comment (Zelldon) [inline atomix/cluster/src/main/java/io/atomix/raft/impl/RaftContext.java]: Do you think we need the max ? So max of current and previous commit? Idk why this was done before. I had hoped that we have the guarantee that this is increasing :sweat_smile:

**Review by ChrisKujawa — APPROVED** [review]
Thanks @deepthidevaki :+1:

File: atomix/cluster/src/main/java/io/atomix/raft/impl/RaftContext.java (hunk 4 above)
Comment (deepthidevaki) [reply atomix/cluster/src/main/java/io/atomix/raft/impl/RaftContext.java]: This is removed code :)

Comment (Zelldon) [reply atomix/cluster/src/main/java/io/atomix/raft/impl/RaftContext.java]: Yeah I know but was not sure whether we need this in the new code? :sweat_smile:

Comment (deepthidevaki) [reply atomix/cluster/src/main/java/io/atomix/raft/impl/RaftContext.java]: Ah.. Sorry. I misread it :smile: It is already checked in `setCommitIndex` https://github.com/zeebe-io/zeebe/blob/2a5a970ec48e97f54aee475ae32757c240dcd03e/atomix/cluster/src/main/java/io/atomix/raft/impl/RaftContext.java#L346

Comment (Zelldon) [reply atomix/cluster/src/main/java/io/atomix/raft/impl/RaftContext.java]: Ah ok this is still a bit different but ok thanks

File: atomix/cluster/src/main/java/io/atomix/raft/roles/LeaderRole.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (MiguelPires) [reply atomix/cluster/src/main/java/io/atomix/raft/roles/LeaderRole.java:96]: No, that probably makes sense then :+1:

**Review by miguelpires — APPROVED** [review]

**Comment by deepthidevaki** [comment]
I guess this doesn't have to be backported? @Zelldon @MiguelPires

**Comment by npepinpe** [comment]
Some questions for back port:

1. Is it fixing a bug?
1. How complicated is it to back port?
1. What are the benefits of back porting if this isn't fixing a bug?

**Comment by Zelldon** [comment]
@npepinpe @deepthidevaki would probably solve https://github.com/zeebe-io/zeebe/issues/3687 for 0.23.X

**Pull request merged**